File: v5_dependency_inversion/python/car_rental_agency2.py
class CarManufactory:
    def rent(self, model: str): pass

class BMW(CarManufactory):
    def rent(self, model: str):
        print(f'BMW rented {model}')

class Mercedes(CarManufactory):
    def rent(self, model: str):
        print(f'Mercedes rented {model}')

class Honda(CarManufactory):
    def rent(self, model: str):
        print(f'Honda rented {model}')
    
class CarRentalAgency2:
    def rent_car(self, manufactory: CarManufactory, model: str):
        manufactory.rent(model)

agency = CarRentalAgency2()
agency.rent_car(BMW(), 'X5')
agency.rent_car(Mercedes(), 'GLE')
agency.rent_car(Honda(), 'Accord')

File: v5_dependency_inversion/java/CarRentalAgency2.java
public class CarRentalAgency2 {

    public static void main(String[] args) {
        CarRentalAgency2 agency = new CarRentalAgency2();
        agency.rentCar(new BMW(), "X5");
        agency.rentCar(new Mercedes(), "GLE");
        agency.rentCar(new Honda(), "Accord");
    }

    public void rentCar(CarManufactory cm, String mode) {
        cm.rent(mode);
    }
    
}

interface CarManufactory {
    public void rent(String model);
}

class BMW implements CarManufactory {
    
    public void rent(String model) {
        System.out.println("BMW rented " + model);
    }
}

class Mercedes implements CarManufactory{

    public void rent(String model) {
        System.out.println("Mercedes rented " + model);
    }
}

class Honda implements CarManufactory {

    public void rent(String model) {
        System.out.println("Honde rented " + model);
    }
}
